src/language-js/print/object.js

```diff
@@ -46,31 +46,27 @@ function printObject(path, options, print) {
     .sort((a, b) => locStart(a) - locStart(b))[0];
 
   const parent = path.getParentNode(0);
+
   const isFlowInterfaceLikeBody =
     isTypeAnnotation &&
     parent &&
     (parent.type === "InterfaceDeclaration" ||
       parent.type === "DeclareInterface" ||
       parent.type === "DeclareClass") &&
     path.getName() === "body";
+
   const shouldBreak =
     n.type === "TSInterfaceBody" ||
     isFlowInterfaceLikeBody ||
     (n.type === "ObjectPattern" &&
-      parent.type !== "FunctionDeclaration" &&
-      parent.type !== "FunctionExpression" &&
-      parent.type !== "ArrowFunctionExpression" &&
-      parent.type !== "ObjectMethod" &&
-      parent.type !== "ClassMethod" &&
-      parent.type !== "ClassPrivateMethod" &&
-      parent.type !== "AssignmentPattern" &&
-      parent.type !== "CatchClause" &&
-      n.properties.some(
-        (property) =>
-          property.value &&
-          (property.value.type === "ObjectPattern" ||
-            property.value.type === "ArrayPattern")
-      )) ||
+      n.properties.filter(
+        ({ value }) =>
+          value &&
+          (value.type === "ObjectPattern" ||
+            value.type === "ArrayPattern" ||
+            (value.type === "AssignmentPattern" &&
+              value.left.type === "ObjectPattern"))
```

Comment:
TODO: should probably check for `value.left.type === "ArrayPattern"` here as well, with a test, for example:

```js
const { first: [inner1, inner2] = [], second: [inner3] = [] } = thing
```

and test a combination like this:

```js
const { first: { inner1 } = {}, second: [inner2, inner3] = [] } = thing
```